File: tests/unit/TaskControllerTests.cpp
#include <QtTest>
#include <QSignalSpy>
#include <QTemporaryDir>
#include "controllers/TaskController.h"
#include "domain/Task.h"
#include "domain/Correction.h"
#include "domain/ChecklistItem.h"

class TaskControllerTests : public QObject
{
    Q_OBJECT

private slots:
    void testDefaults();
    void testAddTask();
    void testAddTaskEmptyName();
    void testRemoveTask();
    void testGetTaskInvalidIndex();
    void testAddItemToTask();
    void testRemoveItemFromTask();
    void testToggleItemInTask();
    void testAddCorrectionToTask();
    void testRemoveCorrectionFromTask();
    void testAddItemToCorrection();
    void testRemoveItemFromCorrection();
    void testToggleItemInCorrection();
    void testSetTaskName();
    void testSetCorrectionName();
    void testSaveAndLoad();
    void testLoadNonExistentFile();
    void testImportFromJsonString();
    void testImportInvalidJson();
    void testAutoSave();
    void testSignals();
};

void TaskControllerTests::testDefaults()
{
    TaskController controller;
    QCOMPARE(controller.taskCount(), 0);
    QVERIFY(!controller.isLoaded());
    QVERIFY(controller.lastError().isEmpty());
    QVERIFY(controller.autoSave());
    QCOMPARE(controller.autoSaveDelay(), 500);
}

void TaskControllerTests::testAddTask()
{
    TaskController controller;
    Task *task = controller.addTask("Mi Tarea");

    QVERIFY(task != nullptr);
    QCOMPARE(controller.taskCount(), 1);
    QCOMPARE(task->name(), QString("Mi Tarea"));
}

void TaskControllerTests::testAddTaskEmptyName()
{
    TaskController controller;
    Task *task = controller.addTask("");

    QVERIFY(task == nullptr);
    QCOMPARE(controller.taskCount(), 0);
    QVERIFY(!controller.lastError().isEmpty());
}

void TaskControllerTests::testRemoveTask()
{
    TaskController controller;
    controller.addTask("Tarea 1");
    controller.addTask("Tarea 2");

    controller.removeTask(0);
    QCOMPARE(controller.taskCount(), 1);
    QCOMPARE(controller.getTask(0)->name(), QString("Tarea 2"));
}

void TaskControllerTests::testGetTaskInvalidIndex()
{
    TaskController controller;
    QVERIFY(controller.getTask(-1) == nullptr);
    QVERIFY(controller.getTask(0) == nullptr);
    QVERIFY(controller.getTask(100) == nullptr);
}

void TaskControllerTests::testAddItemToTask()
{
    TaskController controller;
    controller.addTask("Tarea");

    controller.addItemToTask(0, "Item 1");
    controller.addItemToTask(0, "Item 2");

    Task *task = controller.getTask(0);
    QCOMPARE(task->itemCount(), 2);
    QCOMPARE(task->getItem(0)->text(), QString("Item 1"));
}

void TaskControllerTests::testRemoveItemFromTask()
{
    TaskController controller;
    controller.addTask("Tarea");
    controller.addItemToTask(0, "Item 1");
    controller.addItemToTask(0, "Item 2");

    controller.removeItemFromTask(0, 0);
    QCOMPARE(controller.getTask(0)->itemCount(), 1);
    QCOMPARE(controller.getTask(0)->getItem(0)->text(), QString("Item 2"));
}

void TaskControllerTests::testToggleItemInTask()
{
    TaskController controller;
    controller.addTask("Tarea");
    controller.addItemToTask(0, "Item 1");

    controller.toggleItemInTask(0, 0, true);
    QVERIFY(controller.getTask(0)->getItem(0)->done());

    controller.toggleItemInTask(0, 0, false);
    QVERIFY(!controller.getTask(0)->getItem(0)->done());
}

void TaskControllerTests::testAddCorrectionToTask()
{
    TaskController controller;
    controller.addTask("Tarea");

    Correction *corr = controller.addCorrectionToTask(0, "Corrección 1");
    QVERIFY(corr != nullptr);
    QCOMPARE(controller.getTask(0)->correctionCount(), 1);
    QCOMPARE(corr->name(), QString("Corrección 1"));
}

void TaskControllerTests::testRemoveCorrectionFromTask()
{
    TaskController controller;
    controller.addTask("Tarea");
    controller.addCorrectionToTask(0, "Corr 1");
    controller.addCorrectionToTask(0, "Corr 2");

    controller.removeCorrectionFromTask(0, 0);
    QCOMPARE(controller.getTask(0)->correctionCount(), 1);
    QCOMPARE(controller.getTask(0)->getCorrection(0)->name(), QString("Corr 2"));
}

void TaskControllerTests::testAddItemToCorrection()
{
    TaskController controller;
    controller.addTask("Tarea");
    controller.addCorrectionToTask(0, "Corr 1");

    controller.addItemToCorrection(0, 0, "Item 1");
    QCOMPARE(controller.getTask(0)->getCorrection(0)->itemCount(), 1);
    QCOMPARE(controller.getTask(0)->getCorrection(0)->getItem(0)->text(), QString("Item 1"));
}

void TaskControllerTests::testRemoveItemFromCorrection()
{
    TaskController controller;
    controller.addTask("Tarea");
    controller.addCorrectionToTask(0, "Corr 1");
    controller.addItemToCorrection(0, 0, "Item 1");
    controller.addItemToCorrection(0, 0, "Item 2");

    controller.removeItemFromCorrection(0, 0, 0);
    QCOMPARE(controller.getTask(0)->getCorrection(0)->itemCount(), 1);
    QCOMPARE(controller.getTask(0)->getCorrection(0)->getItem(0)->text(), QString("Item 2"));
}

void TaskControllerTests::testToggleItemInCorrection()
{
    TaskController controller;
    controller.addTask("Tarea");
    controller.addCorrectionToTask(0, "Corr 1");
    controller.addItemToCorrection(0, 0, "Item 1");

    controller.toggleItemInCorrection(0, 0, 0, true);
    QVERIFY(controller.getTask(0)->getCorrection(0)->getItem(0)->done());

    controller.toggleItemInCorrection(0, 0, 0, false);
    QVERIFY(!controller.getTask(0)->getCorrection(0)->getItem(0)->done());
}

void TaskControllerTests::testSetTaskName()
{
    TaskController controller;
    controller.addTask("Original");

    controller.setTaskName(0, "Actualizada");
    QCOMPARE(controller.getTask(0)->name(), QString("Actualizada"));
}

void TaskControllerTests::testSetCorrectionName()
{
    TaskController controller;
    controller.addTask("Tarea");
    controller.addCorrectionToTask(0, "Original");

    controller.setCorrectionName(0, 0, "Actualizada");
    QCOMPARE(controller.getTask(0)->getCorrection(0)->name(), QString("Actualizada"));
}

void TaskControllerTests::testSaveAndLoad()
{
    QTemporaryDir dir;
    QVERIFY(dir.isValid());
    QString filePath = dir.filePath("tasks.json");

    // Save
    {
        TaskController controller(filePath);
        controller.setAutoSave(false);
        controller.addTask("Tarea 1");
        controller.addItemToTask(0, "Item 1");
        controller.addCorrectionToTask(0, "Corr 1");
        controller.addItemToCorrection(0, 0, "Corr Item 1");
        controller.save();
    }

    // Load
    {
        TaskController controller(filePath);
        controller.load();
        QVERIFY(controller.isLoaded());
        QCOMPARE(controller.taskCount(), 1);

        Task *task = controller.getTask(0);
        QCOMPARE(task->name(), QString("Tarea 1"));
        QCOMPARE(task->itemCount(), 1);
        QCOMPARE(task->getItem(0)->text(), QString("Item 1"));
        QCOMPARE(task->correctionCount(), 1);
        QCOMPARE(task->getCorrection(0)->name(), QString("Corr 1"));
        QCOMPARE(task->getCorrection(0)->itemCount(), 1);
        QCOMPARE(task->getCorrection(0)->getItem(0)->text(), QString("Corr Item 1"));
    }
}

void TaskControllerTests::testLoadNonExistentFile()
{
    TaskController controller("/nonexistent/path/tasks.json");
    controller.load();
    QVERIFY(controller.isLoaded());
    QCOMPARE(controller.taskCount(), 0);
}

void TaskControllerTests::testImportFromJsonString()
{
    TaskController controller;
    controller.setAutoSave(false);

    QString json = R"([
        {
            "id": "test-1",
            "name": "Imported Task",
            "date": "2024-01-15",
            "items": [
                {"id": "i1", "text": "Imported Item", "done": false}
            ],
            "corrections": []
        }
    ])";

    controller.importFromJsonString(json);
    QCOMPARE(controller.taskCount(), 1);
    QCOMPARE(controller.getTask(0)->name(), QString("Imported Task"));
    QCOMPARE(controller.getTask(0)->itemCount(), 1);
}

void TaskControllerTests::testImportInvalidJson()
{
    TaskController controller;
    controller.importFromJsonString("not valid json");
    QCOMPARE(controller.taskCount(), 0);
    QVERIFY(!controller.lastError().isEmpty());
}

void TaskControllerTests::testAutoSave()
{
    TaskController controller;
    QVERIFY(controller.autoSave());

    controller.setAutoSave(false);
    QVERIFY(!controller.autoSave());

    controller.setAutoSaveDelay(1000);
    QCOMPARE(controller.autoSaveDelay(), 1000);
}

void TaskControllerTests::testSignals()
{
    TaskController controller;
    QSignalSpy tasksSpy(&controller, &TaskController::tasksChanged);
    QSignalSpy addedSpy(&controller, &TaskController::taskAdded);
    QSignalSpy removedSpy(&controller, &TaskController::taskRemoved);

    controller.addTask("Tarea");
    QCOMPARE(tasksSpy.count(), 1);
    QCOMPARE(addedSpy.count(), 1);

    controller.removeTask(0);
    QCOMPARE(tasksSpy.count(), 2);
    QCOMPARE(removedSpy.count(), 1);
}

QTEST_MAIN(TaskControllerTests)
#include "TaskControllerTests.moc"

File: csharp/tests/Checklist.Core.Tests/TaskControllerTests.cs
using Checklist.Controllers;
using Checklist.Domain;
using Xunit;

namespace Checklist.Core.Tests;

public class TaskControllerTests
{
    [Fact]
    public void Defaults()
    {
        using var controller = new TaskController();
        Assert.Equal(0, controller.TaskCount);
        Assert.False(controller.IsLoaded);
        Assert.Empty(controller.LastError);
        Assert.True(controller.AutoSave);
        Assert.Equal(500, controller.AutoSaveDelay);
    }

    [Fact]
    public void AddTask()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        Task? task = controller.AddTask("Mi Tarea");

        Assert.NotNull(task);
        Assert.Equal(1, controller.TaskCount);
        Assert.Equal("Mi Tarea", task!.Name);
    }

    [Fact]
    public void AddTaskEmptyName()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        Task? task = controller.AddTask("");

        Assert.Null(task);
        Assert.Equal(0, controller.TaskCount);
        Assert.NotEmpty(controller.LastError);
    }

    [Fact]
    public void RemoveTask()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        controller.AddTask("Tarea 1");
        controller.AddTask("Tarea 2");

        controller.RemoveTask(0);
        Assert.Equal(1, controller.TaskCount);
        Assert.Equal("Tarea 2", controller.GetTask(0)?.Name);
    }

    [Fact]
    public void GetTaskInvalidIndex()
    {
        using var controller = new TaskController();
        Assert.Null(controller.GetTask(-1));
        Assert.Null(controller.GetTask(0));
        Assert.Null(controller.GetTask(100));
    }

    [Fact]
    public void AddItemToTask()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        controller.AddTask("Tarea");

        controller.AddItemToTask(0, "Item 1");
        controller.AddItemToTask(0, "Item 2");

        var task = controller.GetTask(0);
        Assert.Equal(2, task?.ItemCount);
        Assert.Equal("Item 1", task?.GetItem(0)?.Text);
    }

    [Fact]
    public void RemoveItemFromTask()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        controller.AddTask("Tarea");
        controller.AddItemToTask(0, "Item 1");
        controller.AddItemToTask(0, "Item 2");

        controller.RemoveItemFromTask(0, 0);
        Assert.Equal(1, controller.GetTask(0)?.ItemCount);
        Assert.Equal("Item 2", controller.GetTask(0)?.GetItem(0)?.Text);
    }

    [Fact]
    public void ToggleItemInTask()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        controller.AddTask("Tarea");
        controller.AddItemToTask(0, "Item 1");

        controller.ToggleItemInTask(0, 0, true);
        Assert.True(controller.GetTask(0)?.GetItem(0)?.Done);

        controller.ToggleItemInTask(0, 0, false);
        Assert.False(controller.GetTask(0)?.GetItem(0)?.Done);
    }

    [Fact]
    public void AddCorrectionToTask()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        controller.AddTask("Tarea");

        var corr = controller.AddCorrectionToTask(0, "Corrección 1");
        Assert.NotNull(corr);
        Assert.Equal(1, controller.GetTask(0)?.CorrectionCount);
        Assert.Equal("Corrección 1", corr?.Name);
    }

    [Fact]
    public void RemoveCorrectionFromTask()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        controller.AddTask("Tarea");
        controller.AddCorrectionToTask(0, "Corr 1");
        controller.AddCorrectionToTask(0, "Corr 2");

        controller.RemoveCorrectionFromTask(0, 0);
        Assert.Equal(1, controller.GetTask(0)?.CorrectionCount);
        Assert.Equal("Corr 2", controller.GetTask(0)?.GetCorrection(0)?.Name);
    }

    [Fact]
    public void AddItemToCorrection()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        controller.AddTask("Tarea");
        controller.AddCorrectionToTask(0, "Corr 1");

        controller.AddItemToCorrection(0, 0, "Item 1");
        Assert.Equal(1, controller.GetTask(0)?.GetCorrection(0)?.ItemCount);
        Assert.Equal("Item 1", controller.GetTask(0)?.GetCorrection(0)?.GetItem(0)?.Text);
    }

    [Fact]
    public void RemoveItemFromCorrection()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        controller.AddTask("Tarea");
        controller.AddCorrectionToTask(0, "Corr 1");
        controller.AddItemToCorrection(0, 0, "Item 1");
        controller.AddItemToCorrection(0, 0, "Item 2");

        controller.RemoveItemFromCorrection(0, 0, 0);
        Assert.Equal(1, controller.GetTask(0)?.GetCorrection(0)?.ItemCount);
        Assert.Equal("Item 2", controller.GetTask(0)?.GetCorrection(0)?.GetItem(0)?.Text);
    }

    [Fact]
    public void SetItemTextInCorrection()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        controller.AddTask("Tarea");
        controller.AddCorrectionToTask(0, "Corr 1");
        controller.AddItemToCorrection(0, 0, "Item 1");

        controller.SetItemTextInCorrection(0, 0, 0, "  Item nuevo  ");
        Assert.Equal("Item nuevo", controller.GetTask(0)?.GetCorrection(0)?.GetItem(0)?.Text);

        controller.Undo();
        Assert.Equal("Item 1", controller.GetTask(0)?.GetCorrection(0)?.GetItem(0)?.Text);

        controller.Redo();
        Assert.Equal("Item nuevo", controller.GetTask(0)?.GetCorrection(0)?.GetItem(0)?.Text);
    }

    [Fact]
    public void ToggleItemInCorrection()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        controller.AddTask("Tarea");
        controller.AddCorrectionToTask(0, "Corr 1");
        controller.AddItemToCorrection(0, 0, "Item 1");

        controller.ToggleItemInCorrection(0, 0, 0, true);
        Assert.True(controller.GetTask(0)?.GetCorrection(0)?.GetItem(0)?.Done);

        controller.ToggleItemInCorrection(0, 0, 0, false);
        Assert.False(controller.GetTask(0)?.GetCorrection(0)?.GetItem(0)?.Done);
    }

    [Fact]
    public void SetTaskName()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        controller.AddTask("Original");

        controller.SetTaskName(0, "Actualizada");
        Assert.Equal("Actualizada", controller.GetTask(0)?.Name);
    }

    [Fact]
    public void SetCorrectionName()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        controller.AddTask("Tarea");
        controller.AddCorrectionToTask(0, "Original");

        controller.SetCorrectionName(0, 0, "Actualizada");
        Assert.Equal("Actualizada", controller.GetTask(0)?.GetCorrection(0)?.Name);
    }

    [Fact]
    public void SaveAndLoad()
    {
        using var dir = new TempDir();
        var filePath = dir.FilePath("tasks.json");

        // Save
        {
            using var controller = new TaskController(filePath);
            controller.AutoSave = false;
            controller.AddTask("Tarea 1");
            controller.AddItemToTask(0, "Item 1");
            controller.AddCorrectionToTask(0, "Corr 1");
            controller.AddItemToCorrection(0, 0, "Corr Item 1");
            Assert.True(controller.Save());
        }

        // Load
        {
            using var controller = new TaskController(filePath);
            controller.AutoSave = false;
            controller.Load();
            Assert.True(controller.IsLoaded);
            Assert.Equal(1, controller.TaskCount);

            var task = controller.GetTask(0);
            Assert.Equal("Tarea 1", task?.Name);
            Assert.Equal(1, task?.ItemCount);
            Assert.Equal("Item 1", task?.GetItem(0)?.Text);
            Assert.Equal(1, task?.CorrectionCount);
            Assert.Equal("Corr 1", task?.GetCorrection(0)?.Name);
            Assert.Equal(1, task?.GetCorrection(0)?.ItemCount);
            Assert.Equal("Corr Item 1", task?.GetCorrection(0)?.GetItem(0)?.Text);
        }
    }

    [Fact]
    public void LoadNonExistentFile()
    {
        using var controller = new TaskController(Path.Combine(Path.GetTempPath(), "nonexistent-" + Guid.NewGuid(), "tasks.json"));
        controller.AutoSave = false;
        controller.Load();
        Assert.True(controller.IsLoaded);
        Assert.Equal(0, controller.TaskCount);
    }

    [Fact]
    public void ImportFromJsonString()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;

        const string json = """
            [
              {
                "id": "test-1",
                "name": "Imported Task",
                "date": "2024-01-15",
                "items": [
                  { "id": "i1", "text": "Imported Item", "done": false }
                ],
                "corrections": []
              }
            ]
            """;

        controller.ImportFromJsonString(json);
        Assert.Equal(1, controller.TaskCount);
        Assert.Equal("Imported Task", controller.GetTask(0)?.Name);
        Assert.Equal(1, controller.GetTask(0)?.ItemCount);
    }

    [Fact]
    public void ImportInvalidJson()
    {
        using var controller = new TaskController();
        controller.ImportFromJsonString("not valid json");
        Assert.Equal(0, controller.TaskCount);
        Assert.NotEmpty(controller.LastError);
    }

    [Fact]
    public void AutoSaveFlags()
    {
        using var controller = new TaskController();
        Assert.True(controller.AutoSave);

        controller.AutoSave = false;
        Assert.False(controller.AutoSave);

        controller.AutoSaveDelay = 1000;
        Assert.Equal(1000, controller.AutoSaveDelay);
    }

    [Fact]
    public void Signals()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;

        var tasksCount = 0;
        var addedCount = 0;
        var removedCount = 0;
        controller.TasksChanged += () => tasksCount++;
        controller.TaskAdded += _ => addedCount++;
        controller.TaskRemoved += _ => removedCount++;

        controller.AddTask("Tarea");
        Assert.Equal(1, tasksCount);
        Assert.Equal(1, addedCount);

        controller.RemoveTask(0);
        Assert.Equal(2, tasksCount);
        Assert.Equal(1, removedCount);
    }

    [Fact]
    public void UndoRedoTaskLifecycle()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;

        controller.AddTask("Tarea");
        Assert.Equal(1, controller.TaskCount);
        Assert.True(controller.CanUndo);
        Assert.False(controller.CanRedo);

        controller.Undo();
        Assert.Equal(0, controller.TaskCount);

        controller.Redo();
        Assert.Equal(1, controller.TaskCount);
        Assert.Equal("Tarea", controller.GetTask(0)?.Name);
    }

    [Fact]
    public void AddTaskWithCategory()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;

        var task = controller.AddTask("Tarea", "Proyecto A");
        Assert.NotNull(task);
        Assert.Equal("Proyecto A", task!.Category);
        Assert.Equal("Proyecto A", controller.GetTask(0)?.Category);
    }

    [Fact]
    public void AddTaskCategoryTrimmed()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;

        var task = controller.AddTask("Tarea", "  Proyecto A  ");
        Assert.Equal("Proyecto A", task?.Category);
    }

    [Fact]
    public void SetTaskCategory()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        controller.AddTask("Tarea");

        controller.SetTaskCategory(0, "Nuevo Tema");
        Assert.Equal("Nuevo Tema", controller.GetTask(0)?.Category);

        controller.SetTaskCategory(0, "");
        Assert.Equal("", controller.GetTask(0)?.Category);
    }

    [Fact]
    public void SetTaskCategoryUndoRedo()
    {
        using var controller = new TaskController();
        controller.AutoSave = false;
        controller.AddTask("Tarea");

        controller.SetTaskCategory(0, "Tema B");
        Assert.True(controller.CanUndo);

        controller.Undo();
        Assert.Equal("", controller.GetTask(0)?.Category);

        controller.Redo();
        Assert.Equal("Tema B", controller.GetTask(0)?.Category);
    }
}